

  
    
  


  




  


  

<!DOCTYPE html>
<html lang="en-us">
  <head>
    
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="generator" content="Hugo 0.27 with theme Tranquilpeak 0.4.1-BETA">
    <title>exprAnalysis package</title>
    <meta name="author" content="Dr. Shirin Glander">
    <meta name="keywords" content="">

    <link rel="icon" href="img/favicon.png">
    

    
    <meta name="description" content="I created the R package exprAnalysis designed to streamline my RNA-seq data analysis pipeline. Below you find the vignette for installation and usage of the package.

This package combines functions from various packages used to analyze and visualize expression data from NGS or expression chips.
 It supports normalized input as e.g. from Cufflinks or expression chip arrays and raw count data from bam file input. It supports mRNA, miRNA, protein or other expression data.">
    <meta property="og:description" content="I created the R package exprAnalysis designed to streamline my RNA-seq data analysis pipeline. Below you find the vignette for installation and usage of the package.

This package combines functions from various packages used to analyze and visualize expression data from NGS or expression chips.
 It supports normalized input as e.g. from Cufflinks or expression chip arrays and raw count data from bam file input. It supports mRNA, miRNA, protein or other expression data.">
    <meta property="og:type" content="blog">
    <meta property="og:title" content="exprAnalysis package">
    <meta property="og:url" content="/2016/09/expranalyis-package/">
    <meta property="og:site_name" content="Shirin&#39;s playgRound">
    <meta name="twitter:card" content="summary">
    <meta name="twitter:title" content="Shirin&#39;s playgRound">
    <meta name="twitter:description" content="I created the R package exprAnalysis designed to streamline my RNA-seq data analysis pipeline. Below you find the vignette for installation and usage of the package.

This package combines functions from various packages used to analyze and visualize expression data from NGS or expression chips.
 It supports normalized input as e.g. from Cufflinks or expression chip arrays and raw count data from bam file input. It supports mRNA, miRNA, protein or other expression data.">
    
      <meta name="twitter:creator" content="@ShirinGlander">
    
    

    
    

    
      <meta property="og:image" content="//www.gravatar.com/avatar/7f7f818e55624edfef8aa93860a1a3d4?s=640">
    

    
      <meta property="og:image" content="/img/rna-seq/microarray/2016/09/28/exprAnalysis_files/figure-markdown_github/heatmaps-1.png">
    
    
    

    

    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/2.1.4/jquery.fancybox.min.css" integrity="sha256-vuXZ9LGmmwtjqFX1F+EKin1ThZMub58gKULUyf0qECk=" crossorigin="anonymous" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/2.1.4/helpers/jquery.fancybox-thumbs.min.css" integrity="sha256-SEa4XYAHihTcEP1f5gARTB2K26Uk8PsndQYHQC1f4jU=" crossorigin="anonymous" />
    
    
    <link rel="stylesheet" href="/css/style-fpbzgxsy0kgmdvyrj5ykkg6ratccrk3gocmaqn4xpcjywmv5dteilzucro4f.min.css" />
    
    

    
      
    
    
  </head>

  <body>
    <div id="blog">
      <header id="header" data-behavior="4">
  <i id="btn-open-sidebar" class="fa fa-lg fa-bars"></i>
  <div class="header-title">
    <a class="header-title-link" href="/">Shirin&#39;s playgRound</a>
  </div>
  
    
      <a class="header-right-picture "
         href="/#about">
    
    
    
      
        <img class="header-picture" src="//www.gravatar.com/avatar/7f7f818e55624edfef8aa93860a1a3d4?s=90" alt="Author&#39;s picture" />
      
    
    </a>
  
</header>

      <nav id="sidebar" data-behavior="4">
  <div class="sidebar-container">
    
      <div class="sidebar-profile">
        <h1 class="sidebar-profile-title">exprAnalysis package</h1>
        <a href="/#about">
          <img class="sidebar-profile-picture" src="//www.gravatar.com/avatar/7f7f818e55624edfef8aa93860a1a3d4?s=110" alt="Author&#39;s picture" />
        </a>
        <h4 class="sidebar-profile-name">Dr. Shirin Glander</h4>
        
          <h5 class="sidebar-profile-bio">Biologist turned Bioinformatician turned Data Scientist</h5>
        
      </div>
    
    <ul class="sidebar-buttons">
      
  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="/">
    
      <i class="sidebar-button-icon fa fa-lg fa-home"></i>
      
      <span class="sidebar-button-desc">Home</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="/categories/">
    
      <i class="sidebar-button-icon fa fa-lg fa-bookmark"></i>
      
      <span class="sidebar-button-desc">Categories</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="/tags/">
    
      <i class="sidebar-button-icon fa fa-lg fa-tags"></i>
      
      <span class="sidebar-button-desc">Tags</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="/archives/">
    
      <i class="sidebar-button-icon fa fa-lg fa-archive"></i>
      
      <span class="sidebar-button-desc">Archives</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="/page/about/">
    
      <i class="sidebar-button-icon fa fa-lg fa-question"></i>
      
      <span class="sidebar-button-desc">About</span>
    </a>
  </li>


    </ul>
    <ul class="sidebar-buttons">
      
  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="https://github.com/ShirinG">
    
      <i class="sidebar-button-icon fa fa-lg fa-github"></i>
      
      <span class="sidebar-button-desc">GitHub</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="https://stackoverflow.com/users/6623620/shirin-glander">
    
      <i class="sidebar-button-icon fa fa-lg fa-stack-overflow"></i>
      
      <span class="sidebar-button-desc">Stack Overflow</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="https://www.r-bloggers.com/">
    
      <i class="sidebar-button-icon fa fa-lg fa-book"></i>
      
      <span class="sidebar-button-desc">R-bloggers</span>
    </a>
  </li>


    </ul>
    <ul class="sidebar-buttons">
      
  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="/index.xml">
    
      <i class="sidebar-button-icon fa fa-lg fa-rss"></i>
      
      <span class="sidebar-button-desc">RSS</span>
    </a>
  </li>


    </ul>
  </div>
</nav>

      

      <div id="main" data-behavior="4"
        class="
               hasCoverMetaOut
               ">
        <article class="post" itemscope itemType="http://schema.org/BlogPosting">
          
          
            <div class="post-header main-content-wrap text-center">
  
    <h1 class="post-title" itemprop="headline">
      exprAnalysis package
    </h1>
  
  
  <div class="postShorten-meta post-meta">
    
      <time itemprop="datePublished" datetime="2016-09-28T00:00:00Z">
        
  September 28, 2016

      </time>
    
    
  
  
    <span>in</span>
    
      <a class="category-link" href="/categories/rna-seq">RNA-seq</a>, 
    
      <a class="category-link" href="/categories/microarray">microarray</a>
    
  


  </div>

</div>
          
          <div class="post-content markdown" itemprop="articleBody">
            <div class="main-content-wrap">
              

<p>I created the R package <strong>exprAnalysis</strong> designed to streamline my RNA-seq data analysis pipeline. Below you find the vignette for installation and usage of the package.</p>

<hr />

<p><br></p>

<p>This package combines functions from various packages used to analyze and visualize expression data from NGS or expression chips.</p>

<ul>
<li>It supports normalized input as e.g. from Cufflinks or expression chip arrays and raw count data from bam file input.</li>
<li>It supports mRNA, miRNA, protein or other expression data.</li>
</ul>

<p>So far, it is implemented for human and mouse data only!</p>

<p>It uses functions from the following packages:</p>

<ul>
<li>AnnotationDbi for annotating gene information</li>
<li>beadarray for importing Illumina expression chip files from GenomeStudio</li>
<li>clusterProfiler and DOSE for functional enrichment analysis</li>
<li>DESeq2 for differential expression analysis of raw count data</li>
<li>GenomicAlignments, GenomicFeatures, Rsamtools for reading bam files</li>
<li>pcaGoPromoter for principle component analysis</li>
<li>limma for differential expression analysis of normalised expression data</li>
<li>pathview for mapping KEGG pathways</li>
<li>gplots for heatmaps</li>
<li>sva for batch correction</li>
<li>WGCNA for coregulatory network determination</li>
</ul>

<hr />

<p><br></p>

<h1 id="prepare-session">Prepare session</h1>

<pre><code class="language-r">library(exprAnalysis)
</code></pre>

<pre><code class="language-r"># update all packages
source.bioc=&quot;https://bioconductor.org/biocLite.R&quot;
source(source.bioc)
biocLite(&quot;BiocUpgrade&quot;)
biocLite(ask=FALSE) 

update.packages()

# make sure the workspace is in pristine condition
rm(list=ls(all=TRUE))

orig_par &lt;- par(no.readonly=T)

options(stringsAsFactors = FALSE)
</code></pre>

<hr />

<p><br></p>

<h2 id="the-package-can-be-installed-via-github">The package can be installed via Github:</h2>

<p>Beware that the vignette is rather large and thus takes a minute to compile. You can also just use this page (it is identical to the vignette).</p>

<pre><code class="language-r"># install package from github
install.packages(&quot;devtools&quot;)
library(devtools)

# either the latest stable release that passed TRAVIS CI check
devtools::install_github(&quot;ShirinG/exprAnalysis&quot;, build_vignettes=TRUE, ref = &quot;stable.version0.1.0&quot;)

# or the development version
devtools::install_github(&quot;ShirinG/exprAnalysis&quot;, build_vignettes=TRUE, ref = &quot;master&quot;)
</code></pre>

<p>There might be problems with installation of some dependency packages (especially Bioconductor packages and WGCNA and its dependencies from CRAN). In order to install them manually:</p>

<pre><code class="language-r">list.of.packages_bioconductor &lt;- c(&quot;arrayQualityMetrics&quot;, &quot;beadarray&quot;, &quot;pcaGoPromoter&quot;, &quot;limma&quot;, &quot;pathview&quot;, &quot;sva&quot;, &quot;GO.db&quot;, &quot;impute&quot;)
list.of.packages_cran &lt;- c(&quot;WGCNA&quot;, &quot;roxygen2&quot;, &quot;testthat&quot;, &quot;gplots&quot;)

new.packages_bioconductor &lt;- list.of.packages_bioconductor[!(list.of.packages_bioconductor %in% installed.packages()[,&quot;Package&quot;])]
new.packages_cran &lt;- list.of.packages_cran[!(list.of.packages_cran %in% installed.packages()[,&quot;Package&quot;])]

# CRAN
if(length(new.packages_cran)&gt;0) install.packages(new.packages_cran)

# Bioconductor
if(length(new.packages_bioconductor)&gt;0) {
  source(&quot;https://bioconductor.org/biocLite.R&quot;)
  biocLite(new.packages_bioconductor)
}
</code></pre>

<p>Load the library:</p>

<pre><code class="language-r">library(&quot;exprAnalysis&quot;)

# To view the vignette if you built it with the package:

vignette(&quot;exprAnalysis&quot;, package=&quot;exprAnalysis&quot;)
vignette(&quot;CummeRbund&quot;, package=&quot;exprAnalysis&quot;)

browseVignettes(&quot;exprAnalysis&quot;)

enableWGCNAThreads()
</code></pre>

<hr />

<p><br></p>

<h1 id="functions">Functions</h1>

<p>See package help (?functionname, ?data) for detailed descriptions of functions and example datasets.</p>

<hr />

<p><br></p>

<h1 id="input-data">Input data</h1>

<p>Takes an expression data matrix containing numeric values as expression measures (e.g. read count data, FPKM values, Illumina expression data).</p>

<ul>
<li>Rownames should be gene or isoform identifiers (e.g. gene names)</li>
<li>Colnames should be sample IDs (sample names)</li>
</ul>

<p>This package contains two example matrices of randomly generated expression data as raw read counts (called <strong>&ldquo;countmatrix&rdquo;</strong>) and as normalized read counts (called <strong>&ldquo;expmatrix&rdquo;</strong>).</p>

<pre><code class="language-r">data(&quot;countmatrix&quot;)
data(&quot;expmatrix&quot;)
</code></pre>

<hr />

<p><br></p>

<h2 id="starting-from-fpkm-data">Starting from FPKM data</h2>

<p>If you have FPKM data (e.g. quantile normalized) from Cufflinks, treat the data such as the <strong>expmatrix</strong> example.</p>

<p><br></p>

<h3 id="cuffdiff">Cuffdiff</h3>

<p>See CummeRbund vignette included among the vignettes of this package <strong>CummeRbundVignette.html</strong>.</p>

<hr />

<p><br></p>

<h2 id="starting-from-count-data">Starting from count data</h2>

<p>If you want to do count data analysis, you can either produce a count matrix (e.g. with HTSeq) and proceed to DESeq2 analysis or you can produce the count table directly from the bam files and then proceed to DESeq2.</p>

<p><strong>read_bam_to_countmatrix()</strong> returns a DESeq data set that can directly be used with the DEseq2 pipeline. Or you can manually load the count matrix that was saved by <strong>read_bam_to_countmatrix()</strong> and then go to DESeqDataFrameFromMatrix().</p>

<p>In <strong>read_bam_to_countmatrix()</strong> fragments will be counted only once to each gene, even if they overlap multiple exons of a gene which may themselves be overlapping.</p>

<p><strong>read_bam_to_countmatrix()</strong> uses the packages Rsamtools, GenomicAlignments, GenomicFeatures, Biobase and SummarizedExperiment.</p>

<pre><code class="language-r"># Locate alignment files
dir &lt;- getwd()
filenames &lt;- fileslist[grep(&quot;.*sorted.bam$&quot;, list.files(dir))]

# Create a sample table

samplename &lt;- sub(&quot;_accepted_hits.sorted.bam&quot;, &quot;&quot;, filenames)
design &lt;- c(&quot;Treatment&quot;, &quot;Control&quot;)
sampleid &lt;- sub(&quot;Sample&quot;, &quot;&quot;, samplename)

sampleTable &lt;- data.frame(row.names = samplename, sampleid = sampleid, filenames = filenames, colData = design, stringsAsFactors = F)

#         sampleid                         filenames    colData
#Sample1         1  Sample1_accepted_hits.sorted.bam  Treatment
#Sample2         2  Sample2_accepted_hits.sorted.bam    Control

data &lt;- read_bam_to_countmatrix(sampleTable, gtffile = &quot;Homo_sapiens.GRCh38.83.gtf&quot;, projectfolder = getwd(), outPrefix=&quot;Test&quot;)

countmatrix &lt;- SummarizedExperiment::assay(data)
</code></pre>

<p><br></p>

<h3 id="continue-count-data-analysis-with-deseq2">Continue count data analysis with DESeq2</h3>

<p>See DESeq2 Vignette for additional details.</p>

<h4 id="if-you-do-not-directly-proceed-from-read-bam-to-countmatrix">If you do not directly proceed from read_bam_to_countmatrix():</h4>

<ul>
<li>read in saved count matrix</li>
<li>define experimental design</li>
<li>convert to DESeq data set</li>
</ul>

<pre><code class="language-r">countmatrix &lt;-   read.table(file.path(projectfolder, &quot;Countdata&quot;, outPrefix, paste0(outPrefix, &quot;_Summarize_overlaps_count_matrix.txt&quot;)), header=T, sep = &quot;\t&quot;)
</code></pre>

<pre><code class="language-r">design &lt;- gsub(&quot;(.*)(_[0-9])&quot;, &quot;\\1&quot;, colnames(countmatrix))
ExpDesign &lt;- data.frame(row.names=colnames(countmatrix), treatment = design)

data &lt;- DESeq2::DESeqDataSetFromMatrix(countData = countmatrix, colData=ExpDesign, design=~treatment)
</code></pre>

<p><br></p>

<h3 id="deseq2">DESeq2</h3>

<ul>
<li>optional, but recommended: remove genes with zero counts over all samples</li>
<li>run DESeq</li>
<li>Extracting transformed values</li>
</ul>

<p>Note: the rlog transformation is provided for applications other than differential testing. For differential testing we recommend the DESeq function applied to raw counts, as described later in this workflow, which also takes into account the dependence of the variance of counts on the mean value during the dispersion estimation step.</p>

<pre><code class="language-r">data &lt;- data[rowSums(DESeq2::counts(data)) &gt; 1, ]

data_DESeq &lt;- DESeq2::DESeq(data)

expmatrix_DESeq &lt;- DESeq2::rlog(data_DESeq, fitType=&quot;local&quot;)
expmatrix &lt;- SummarizedExperiment::assay(expmatrix_DESeq)
</code></pre>

<p><br></p>

<h4 id="dispersion-plot">Dispersion plot</h4>

<pre><code class="language-r">DESeq2::plotDispEsts(data_DESeq, main=&quot;Dispersion Estimates&quot;)
</code></pre>

<hr />

<p><br></p>

<h2 id="starting-from-affymetrix-expression-chips">Starting from Affymetrix expression chips</h2>

<p>If you have CEL files, start with the following code to produce the expression matrix and then treat it like the <strong>expmatrix</strong> example:</p>

<ul>
<li>Read in the data and create an expression, using RMA for example</li>
</ul>

<p>Currently the rma function implements RMA in the following manner 1. Probe specific correction of the PM probes using a model based on observed intensity being the sum of signal and noise 2. Normalization of corrected PM probes using quantile normalization (Bolstad et al., 2003) 3. Calculation of Expression measure using median polish.</p>

<pre><code class="language-r">library(affy)

# Choose directory containing CEL files
dir &lt;- getwd()

# check, that you have the correct directory
celfiles &lt;- file.path(dir, list.files(dir, recursive = TRUE)[grep(&quot;.*CEL$&quot;, list.files(dir, recursive = TRUE), ignore.case = TRUE)])

affydata &lt;- ReadAffy(filenames=celfiles)
affy::MAplot(affydata,plot.method=&quot;smoothScatter&quot;)
eset &lt;- affy::rma(affydata)

# If affy doesn' recognize the chip type, try the oligo package:
library(oligo)
rawData &lt;- read.celfiles(celfiles)
oligo::MAplot(rawData, plotFun=smoothScatter)
eset &lt;- backgroundCorrect(rawData, method=&quot;rma&quot;)

quality_control_plots(eset, groupColumn=c())

# Optional: Remove bad quality probes/ genes and/ or samples

write.exprs(eset, file=&quot;eset.txt&quot;)

library(made4)
overview(eset)

# RNA degradation plots
deg &lt;- AffyRNAdeg(affydata)
plotAffyRNAdeg(deg)


expmatrix &lt;- exprs(eset)
</code></pre>

<hr />

<p><br></p>

<h2 id="starting-from-illumina-expression-chips">Starting from Illumina expression chips</h2>

<p>If you have Illumina intensity data, load it into GenomeStudio first:</p>

<ul>
<li>Open New Project -&gt; Gene Expression</li>
<li>Choose Assay Type -&gt; Next</li>
<li>Choose Project Repo and name the project -&gt; Next</li>
<li>On the left side, mark all folders containing data, select &ldquo;All&rdquo; and import them into the right window -&gt; Next</li>
<li>Name Group Set, on the left side, mark all folders containing data, select &ldquo;All&rdquo; and import them into the right window (optionally, you can directly create groups here) -&gt; Next</li>
<li>For now, don&rsquo;t use normalisation and don&rsquo; substract background</li>
<li>Choose appropriate Content Descriptor (e.g. HumanHT-12_V4_0_R1_15002873_B.bgx) -&gt; Finish</li>
<li>Export SampleProbeProfile.txt: Choose columns AVG_Signal, Detection Pval, BEAD_STDERR and Avg_NBEADS</li>
<li>Export ControlProbeProfile.txt: Choose columns AVG_Signal and Detection Pval</li>
<li>You also need the Samplesheet.csv</li>
</ul>

<p><strong>read_Illumina()</strong> and <strong>normalise_eset()</strong> uses the R package beadarray.</p>

<pre><code class="language-r">dataFile      &lt;- file.path(&quot;SampleProbeProfile.txt&quot;)
qcFile        &lt;- file.path(&quot;ControlProbeProfile.txt&quot;)
sampleSheet   &lt;- file.path(&quot;samplesheet.csv&quot;)

# define Illumina Expression Array: &quot;HumanHT-12 v?&quot;, &quot;MouseWG-6 v?&quot;, &quot;MouseRef-8 v?&quot;
expressionchipType &lt;- &quot;HumanHT-12 v4&quot;

eset &lt;- read_Illumina(dataFile, qcFile, sampleSheet, expressionchipType, ProbeID = &quot;PROBE_ID&quot;, skip = 0, controlID=&quot;ProbeID&quot;, qc.skip = 0, method_norm =&quot;none&quot;, transform=&quot;log2&quot;)

quality_control_plots(eset)

# Optional: Remove bad quality probes/ genes and/ or samples

eset &lt;- normalise_eset(eset)

expmatrix &lt;- exprs(eset)
</code></pre>

<hr />

<p><br></p>

<h1 id="batch-correction">Batch correction</h1>

<p><strong>batch_removal()</strong> uses the R package sva.</p>

<pre><code class="language-r">pheno &lt;- data.frame(sample=c(1:16), treatment=sub(&quot;(.*)(_[0-9])&quot;, &quot;\\1&quot;, colnames(expmatrix)), batch=ifelse(grepl(&quot;Ctrl&quot;, colnames(expmatrix)) == TRUE, &quot;1&quot;, ifelse(grepl(&quot;ActLPS&quot;, colnames(expmatrix)) == TRUE, &quot;1&quot;, &quot;2&quot;)), row.names = colnames(expmatrix))

expmatrix &lt;- batch_removal(expmatrix, pheno)
</code></pre>

<hr />

<p><br></p>

<h1 id="exploratory-analysis-of-all-genes">Exploratory analysis of all genes</h1>

<h2 id="variance-vs-mean-gene-expression-across-samples">Variance vs mean gene expression across samples</h2>

<p>Plots variance against mean gene expression across samples and calculates the correlation of a linear regression model.</p>

<p><strong>var_vs_mean()</strong> uses the R package matrixStats.</p>

<pre><code class="language-r">var_vs_mean(countmatrix)
</code></pre>

<p><img src="/img/rna-seq/microarray/2016/09/28/exprAnalysis_files/figure-markdown_github/var_vs_mean-1.png" alt="" /></p>

<pre><code>## 
##  Pearson's product-moment correlation
## 
## data:  log2(dispersion[, 1] + 1) and log2(dispersion[, 2] + 1)
## t = 281.2, df = 9998, p-value &lt; 2.2e-16
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  0.9399669 0.9443684
## sample estimates:
##       cor 
## 0.9422083
</code></pre>

<pre><code class="language-r"># var_vs_mean(expmatrix)
</code></pre>

<hr />

<p><br></p>

<h2 id="intersample-variances">Intersample variances</h2>

<pre><code class="language-r">library(corrgram)

Ctrl_cor &lt;- expmatrix[,grep(&quot;Ctrl&quot;, colnames(expmatrix))]

corrgram::corrgram(Ctrl_cor, order=TRUE, lower.panel=corrgram::panel.pie,
         upper.panel=corrgram::panel.pts, text.panel=corrgram::panel.txt,
         main=&quot;Correlogram of controls&quot;)
</code></pre>

<hr />

<p><br></p>

<h2 id="principle-component-analysis">Principle Component Analysis</h2>

<p>Uses functions from the R package pcaGoPromoter.</p>

<p>You can only plot the principle components using:</p>

<pre><code class="language-r">groups &lt;- as.factor(c(rep(&quot;Ctrl&quot;,4), rep(&quot;TolLPS&quot;,4), rep(&quot;TolS100A8&quot;,4), rep(&quot;ActLPS&quot;,4)))

pca_plot(expmatrix, groups)
</code></pre>

<p><img src="/img/rna-seq/microarray/2016/09/28/exprAnalysis_files/figure-markdown_github/pca_plot-1.png" alt="" /></p>

<p><br></p>

<p>Or you can plot the principle components and calculate TF and GO term enrichments of genes (defaults to top 2.5%) with highest and lowest loadings. With this function, the ouput files are directly saved to .pdf and .txt (by default to working directory).</p>

<pre><code class="language-r">pca_plot_enrich(expmatrix, groups)
</code></pre>

<hr />

<p><br></p>

<h2 id="heatmaps">Heatmaps</h2>

<p><strong>heatmaps()</strong> uses the R package gplots.</p>

<p>Here, of the 30 most highly expressed genes.</p>

<pre><code class="language-r">select &lt;- order(rowMeans(expmatrix),decreasing=TRUE)[1:30]

heatmaps(expmatrix[select,], samplecols = rep(c(&quot;#E41A1C&quot;, &quot;#377EB8&quot;, &quot;#4DAF4A&quot;, &quot;#984EA3&quot;), each=4))
</code></pre>

<p><img src="/img/rna-seq/microarray/2016/09/28/exprAnalysis_files/figure-markdown_github/heatmaps-1.png" alt="" /></p>

<pre><code class="language-r"># Heatmap function from DESeq2, using pheatmap:

library(pheatmap)

sampleDists &lt;- dist(t(expmatrix))
sampleDistMatrix &lt;- as.matrix(sampleDists)
rownames(sampleDistMatrix) &lt;- paste(expmatrix_DESeq$treatment)
colnames(sampleDistMatrix) &lt;- NULL
colors &lt;- grDevices::colorRampPalette( rev(RColorBrewer::brewer.pal(9, &quot;Blues&quot;)) )(255)
pheatmap::pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

df &lt;- data.frame(treatment = SummarizedExperiment::colData(data_DESeq)[,c(&quot;treatment&quot;)], row.names = rownames(SummarizedExperiment::colData(data_DESeq)))
pheatmap::pheatmap(expmatrix[select,], cluster_rows=TRUE, show_rownames=TRUE, cluster_cols=TRUE, annotation_col=df)
</code></pre>

<pre><code class="language-r"># Interactive heatmap

#devtools::install_github('talgalili/heatmaply')
library(heatmaply)

# sample correlation
hm &lt;- heatmapr(cor(expmatrix[select,]), k_row = 4, k_col = 4)

# gene correlation
hm &lt;- heatmapr(cor(t(expmatrix[select,])), k_row = NULL, k_col = NULL)

heatmaply(hm)

#Hover over heatmap to see individual values and sample/ gene IDs
</code></pre>

<hr />

<p><br></p>

<h2 id="wgcna">WGCNA</h2>

<p>Check <a href="https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/">WGCNA page</a> for detailed description of the WGCNA package.</p>

<h3 id="hierarchical-clustering-and-outlier-detection">Hierarchical Clustering and outlier detection</h3>

<p>Uses adjacency matrix function from the R package WGCNA and hierarchical clustering from the R package flashClust.</p>

<pre><code class="language-r">datTraits &lt;- data.frame(Ctrl = c(rep(1, 4), rep(0,12)), TolPS = c(rep(0, 4), rep(1, 4),rep(0, 8)), TolS100A8 = c(rep(0, 8), rep(1, 4), rep(0, 4)), ActLPS = c(rep(0, 12),rep(1, 4)), Tol = c(rep(0, 4), rep(1, 8), rep(0, 4)), ExPhenotype = c(stats::rnorm(4, 10, 1),stats::rnorm(8, 25, 1),stats::rnorm(4, 50, 1)), row.names = colnames(expmatrix))

datExpr &lt;- wgcna_sample_dendrogram(expmatrix, datTraits)
</code></pre>

<pre><code>## 
</code></pre>

<p><img src="/img/rna-seq/microarray/2016/09/28/exprAnalysis_files/figure-markdown_github/wgcna_sample_dendrogram-1.png" alt="" /></p>

<pre><code>##  Flagging genes and samples with too many missing values...
##   ..step 1
## All genes are okay!
## All samples are okay!
</code></pre>

<pre><code class="language-r"># Optional: Remove outlier samples and repeats: All genes flagged for removal are saved to the object &quot;remove_genes&quot;
#head(remove_genes)
</code></pre>

<p><br></p>

<h3 id="choosing-a-soft-threshold">Choosing a Soft Threshold</h3>

<p>Ideally, pick a SFT with R^2 &gt; 0.9. In this example, this threshold was not reached, so I pick the highest SFT: 30.</p>

<pre><code class="language-r"># Choose a set of soft thresholding powers
powers=c(1:30)

# choose power based on SFT criterion
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5, networkType =&quot;signed&quot;, corFnc= &quot;bicor&quot;, corOptions=list(maxPOutliers=0.1))

# Plot the results:
tiff(filename= &quot;SFT.tiff&quot;, width = 8000 , height = 4000, res=600, compression = &quot;lzw&quot;)
par(mfrow=c(1,2))
# SFT index as a function of different powers
plot(sft$fitIndices[,1],-sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab=&quot;Soft Threshold (power)&quot;,ylab=&quot;SFT, signed R^2&quot;,type=&quot;n&quot;,main=paste(&quot;Scale independence&quot;))
text(sft$fitIndices[,1],-sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,col=&quot;red&quot;)
# this line corresponds to using an R^2 cut-off of h
abline(h=0.90,col=&quot;red&quot;)
# Mean connectivity as a function of different powers
plot(sft$fitIndices[,1],sft$fitIndices[,5],type=&quot;n&quot;,
     xlab=&quot;Soft Threshold (power)&quot;,ylab=&quot;Mean Connectivity&quot;,main=paste(&quot;Mean connectivity&quot;))
text(sft$fitIndices[,1],sft$fitIndices[,5],labels=powers,col=&quot;red&quot;)
dev.off()
</code></pre>

<p><br></p>

<h3 id="automatic-module-detection-via-dynamic-tree-cutting">Automatic module detection via dynamic tree cutting</h3>

<pre><code class="language-r">softpower = 30
mergingThresh = 0.25
net = blockwiseModules(datExpr, checkMissingData = TRUE, corType = &quot;bicor&quot;, # or pearson
                       maxBlockSize = 10000, networkType = &quot;signed&quot;, power = softpower,
                       minModuleSize = 30, maxPOutliers = 0.1, mergeCutHeight = mergingThresh,
                       numericLabels = TRUE, saveTOMs = TRUE, randomSeed = 12345,
                       pamRespectsDendro = FALSE, saveTOMFileBase = &quot;TESTexprsTOM&quot;)
str(net)
</code></pre>

<pre><code class="language-r">moduleColors &lt;- wgcna_plotDendroAndColors(datExpr, datTraits, net)
</code></pre>

<p><br></p>

<pre><code class="language-r"># Recalculate MEs with color labels
MEs0 = moduleEigengenes(datExpr, moduleColors, softPower = softpower)$eigengenes
# Reorder given (eigen-)vectors such that similar ones (as measured by correlation) are next to each other:
MEs = orderMEs(MEs0)
rownames(MEs) &lt;- rownames(datExpr)
write.table(MEs, &quot;Automatic_ModuleEigengenes_numberLabel.txt&quot;, row.names=T, quote=F, sep=&quot;\t&quot;)

# calculate the module membership values (aka. module eigengene based connectivity kME):
  datKME &lt;- WGCNA::signedKME(datExpr, MEs)  # equals geneModuleMembership
  colnames(datKME) &lt;- sub(&quot;kME&quot;, &quot;MM.&quot;, colnames(datKME))

wgcna_modulememberships(datExpr, MEs, datKME, moduleColors)

wgcna_heatmaps(MEs, datExpr, datTraits)
</code></pre>

<p><img src="wgcnaHeatmap.png" alt="Module trait correlation heatmap" /></p>

<p><br></p>

<h3 id="output-file-for-gene-ontology-analysis">Output file for gene ontology analysis</h3>

<p><strong>geneAnnotations()</strong> uses the R package AnnotationDbi.</p>

<pre><code class="language-r"># Annotate
genes = colnames(datExpr)
datExpr_anno &lt;- geneAnnotations(input=genes, keys=genes, column=c(&quot;ENTREZID&quot;, &quot;ENSEMBL&quot;), keytype=&quot;SYMBOL&quot;, organism = &quot;human&quot;)

# Correlations of genes and traits
traitsCor=cor(datExpr,datTraits,use=&quot;p&quot;)
colnames(traitsCor)=paste(&quot;cor&quot;,colnames(traitsCor),sep=&quot;.&quot;)

dataOutput=data.frame(datExpr_anno ,moduleColors, datKME, traitsCor)

write.table(dataOutput,&quot;AllGenesResults.txt&quot;,row.names=F,sep=&quot;\t&quot;)
</code></pre>

<p><br></p>

<h3 id="module-eigengene-expression-plots">Module eigengene expression plots</h3>

<pre><code class="language-r">data(MEs)
</code></pre>

<pre><code class="language-r">layout(matrix(c(1:length(colnames(MEs)),12,rep(13,3)), ncol=3, byrow=TRUE), heights=c(rep(3,4), 0.5))

for (color in gsub(&quot;ME&quot;, &quot;&quot;, colnames(MEs))){
  module_eigengene_plot(groups, MEs, color)
}
plot.new()

par(mai=c(0,0,0,0))
plot.new()
legend(&quot;bottom&quot;, c(&quot;mean module eigengenes&quot;, &quot;replicate module eigengene values&quot;), xpd = TRUE, horiz = TRUE, inset = c(0,0), bty = &quot;n&quot;, col = c(&quot;grey&quot;, &quot;black&quot;), pch = 16, cex = 2)
</code></pre>

<p><img src="/img/rna-seq/microarray/2016/09/28/exprAnalysis_files/figure-markdown_github/module_eigengene_plot-1.png" alt="" /></p>

<hr />

<p><br></p>

<h1 id="differential-expression-analysis">Differential expression analysis</h1>

<h2 id="de-analyis-using-the-r-package-limma">DE analyis using the R package limma</h2>

<p>For normalized expression data (FPKM from Cufflinks or intensities from expression chip arrays).</p>

<pre><code class="language-r">design &lt;- as.matrix(data.frame(Ctrl = c(rep(1, 4), rep(0,12)), TolLPS = c(rep(0, 4), rep(1, 4),
              rep(0, 8)), TollMRP8 = c(rep(0, 8), rep(1, 4), rep(0, 4)), ActLPS = c(rep(0, 12),
              rep(1, 4)), row.names = colnames(expmatrix)))
groupcomparisons=c(&quot;TolLPS-Ctrl&quot;, &quot;TollMRP8-Ctrl&quot;, &quot;ActLPS-Ctrl&quot;)
DEgenes_all &lt;- diff_limma_all(expmatrix, design, groupcomparisons)

comparison=&quot;TolLPS-Ctrl&quot;
Allgenes_limma_pw &lt;- diff_limma_pw_unfiltered(expmatrix, design, comparison)

DEgenes_pw &lt;- Allgenes_limma_pw[which(abs(Allgenes_limma_pw$logFC) &gt; log2(1.5) &amp; Allgenes_limma_pw$adj.P.Val &lt; 0.05),]
</code></pre>

<p>Example output files from diff_limma functions have been saved to the package data and can be accessed via:</p>

<pre><code class="language-r">data(&quot;DEgenes_pw&quot;)
data(&quot;Allgenes_limma_pw&quot;)


# Export tables to Latex
library(xtable)

italic &lt;- function(x){
paste0('{\\emph{ ', x, '}}')
}

print(xtable(DEgenes_pw[1:10,], 
       digits = 2,
       caption = c(&quot;\\tt Long caption&quot;, &quot;Short caption&quot;),
       label = &quot;tab:testtable&quot;,
       auto = TRUE),
      caption.placement = &quot;top&quot;,
      sanitize.rownames.function = italic,
      booktabs = TRUE,
      floating = TRUE, 
      latex.environments = &quot;center&quot;,
      include.rownames = TRUE,
      scalebox = 0.8,
      tabular.environment = &quot;tabularx&quot;, width = &quot;\\textwidth&quot;)

# or to html
print(xtable(DEgenes_pw[1:10,], 
       digits = 2,
       caption = c(&quot;Long caption&quot;, &quot;Short caption&quot;),
       label = &quot;tab:testtable&quot;,
       auto = TRUE), type = &quot;html&quot;, caption.placement = &quot;top&quot;)
</code></pre>

<!-- html table generated in R 3.3.0 by xtable 1.8-2 package -->

<!-- Wed Jun 01 08:44:34 2016 -->

<table border="1">
<caption align="top">
Long caption
</caption>
<tr>
<th>
</th>
<th>
logFC
</th>
<th>
CI.L
</th>
<th>
CI.R
</th>
<th>
AveExpr
</th>
<th>
t
</th>
<th>
P.Value
</th>
<th>
adj.P.Val
</th>
<th>
B
</th>
<th>
ENTREZID
</th>
<th>
ENSEMBL
</th>
</tr>
<tr>
<td>
HCAR2
</td>
<td align="right">
-1.60
</td>
<td align="right">
-1.78
</td>
<td align="right">
-1.41
</td>
<td align="right">
5.17
</td>
<td align="right">
-18.38
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
16.91
</td>
<td>
338442
</td>
<td>
ENSG00000182782
</td>
</tr>
<tr>
<td>
LOC100996342
</td>
<td align="right">
2.54
</td>
<td align="right">
2.18
</td>
<td align="right">
2.90
</td>
<td align="right">
6.89
</td>
<td align="right">
15.03
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
14.33
</td>
<td>
100996342
</td>
<td>
ENSG00000235478
</td>
</tr>
<tr>
<td>
INPP4B
</td>
<td align="right">
-1.54
</td>
<td align="right">
-1.77
</td>
<td align="right">
-1.31
</td>
<td align="right">
4.80
</td>
<td align="right">
-14.19
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
13.58
</td>
<td>
8821
</td>
<td>
ENSG00000109452
</td>
</tr>
<tr>
<td>
LOC100505549
</td>
<td align="right">
-1.26
</td>
<td align="right">
-1.45
</td>
<td align="right">
-1.06
</td>
<td align="right">
3.11
</td>
<td align="right">
-13.91
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
13.32
</td>
<td>
100505549
</td>
<td>
</td>
</tr>
<tr>
<td>
FCGBP
</td>
<td align="right">
-2.06
</td>
<td align="right">
-2.38
</td>
<td align="right">
-1.75
</td>
<td align="right">
3.39
</td>
<td align="right">
-13.87
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
13.28
</td>
<td>
8857
</td>
<td>
ENSG00000275395
</td>
</tr>
<tr>
<td>
ZSCAN31
</td>
<td align="right">
-1.40
</td>
<td align="right">
-1.63
</td>
<td align="right">
-1.18
</td>
<td align="right">
3.44
</td>
<td align="right">
-13.33
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
12.76
</td>
<td>
64288
</td>
<td>
ENSG00000235109
</td>
</tr>
<tr>
<td>
MIR3665
</td>
<td align="right">
-1.32
</td>
<td align="right">
-1.53
</td>
<td align="right">
-1.11
</td>
<td align="right">
2.69
</td>
<td align="right">
-13.27
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
12.70
</td>
<td>
100500861
</td>
<td>
ENSG00000266325
</td>
</tr>
<tr>
<td>
PTH2R
</td>
<td align="right">
-1.62
</td>
<td align="right">
-1.90
</td>
<td align="right">
-1.35
</td>
<td align="right">
3.69
</td>
<td align="right">
-12.56
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
11.98
</td>
<td>
5746
</td>
<td>
ENSG00000144407
</td>
</tr>
<tr>
<td>
GCLC
</td>
<td align="right">
-2.42
</td>
<td align="right">
-2.83
</td>
<td align="right">
-2.00
</td>
<td align="right">
3.13
</td>
<td align="right">
-12.27
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
11.66
</td>
<td>
2729
</td>
<td>
ENSG00000001084
</td>
</tr>
<tr>
<td>
SOX30
</td>
<td align="right">
2.54
</td>
<td align="right">
2.10
</td>
<td align="right">
2.98
</td>
<td align="right">
8.57
</td>
<td align="right">
12.19
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
11.59
</td>
<td>
11063
</td>
<td>
ENSG00000039600
</td>
</tr>
</table>

<hr />

<p><br></p>

<h2 id="de-analyis-using-deseq2">DE analyis using DESeq2</h2>

<p>For raw read count data.</p>

<h5 id="contrast-de-groups">contrast DE groups:</h5>

<ul>
<li>lfc = treatment &gt; Ctrl, - lfc = treatment &lt; Ctrl p-value &amp; p.adjust values of NA indicate outliers detected by Cook&rsquo;s distance NA only for p.adjust means the gene is filtered by automatic independent filtering for having a low mean normalized count</li>
</ul>

<p><br></p>

<pre><code class="language-r">library(DESeq2)
library(ggplot2)
library(ggrepel)

# find possible contrasts with
DESeq2::resultsNames(data_DESeq)

res &lt;- DESeq2::results(data_DESeq, contrast=list(&quot;treatmentActLPS&quot;, &quot;treatmentCtrl&quot;), cooksCutoff = 0.99, independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = &quot;BH&quot;)

# order results table by the smallest adjusted p value:
res &lt;- res[order(res$padj),]

results = as.data.frame(dplyr::mutate(as.data.frame(res), sig=ifelse(res$padj&lt;0.05, &quot;FDR&lt;0.05&quot;, &quot;Not Sig&quot;)), row.names=rownames(res))

DEgenes_DESeq &lt;- results[which(abs(results$log2FoldChange) &gt; log2(1.5) &amp; results$padj &lt; 0.05),]

p = ggplot2::ggplot(results, ggplot2::aes(log2FoldChange, -log10(pvalue))) +
  ggplot2::geom_point(ggplot2::aes(col=sig)) +
  ggplot2::scale_color_manual(values=c(&quot;red&quot;, &quot;black&quot;)) +
  ggplot2::ggtitle(&quot;Volcano Plot of DESeq2 analysis&quot;)

p+ggrepel::geom_text_repel(data=results[1:10, ], ggplot2::aes(label=rownames(results[1:10, ])))

# If there aren't too many DE genes:
#p+geom_text_repel(data=dplyr::filter(results, padj&lt;0.05), aes(label=rownames(results[1:10, ])))
</code></pre>

<p><br></p>

<h4 id="ma-plot">MA-plot</h4>

<pre><code class="language-r">DESeq2::plotMA(res, main=&quot;MA Plot&quot;, ylim=c(-2,2))
</code></pre>

<p><br></p>

<h4 id="plotcounts-which-normalizes-counts-by-sequencing-depth-and-adds-a-pseudocount-of-1-2-to-allow-for-log-scale-plotting">plotCounts, which normalizes counts by sequencing depth and adds a pseudocount of <sup>1</sup>&frasl;<sub>2</sub> to allow for log scale plotting</h4>

<pre><code class="language-r">par(mfrow=c(1,3))

for (i in 1:3){
  gene &lt;- rownames(res)[i]
  main = gene
  DESeq2::plotCounts(data_DESeq, gene=gene, intgroup=&quot;treatment&quot;, main = main)
}
</code></pre>

<hr />

<p><br></p>

<h2 id="gene-annotations">Gene annotations</h2>

<p>Can be used to add e.g. ENTREZ ID, ENSEMBL ID, etc. to gene name.</p>

<pre><code class="language-r">DEgenes_pw &lt;- geneAnnotations(input=DEgenes_pw, keys=row.names(DEgenes_pw), column=c(&quot;ENTREZID&quot;, &quot;ENSEMBL&quot;), keytype=&quot;SYMBOL&quot;, organism = &quot;human&quot;)
</code></pre>

<pre><code class="language-r">DEgenes_DESeq &lt;- geneAnnotations(input=DEgenes_DESeq, keys=row.names(DEgenes_DESeq), column=c(&quot;ENTREZID&quot;, &quot;ENSEMBL&quot;), keytype=&quot;SYMBOL&quot;)
</code></pre>

<hr />

<p><br></p>

<h2 id="enrichment-analysis-using-clusterpofiler">Enrichment Analysis using clusterPofiler</h2>

<pre><code class="language-r">library(clusterProfiler)
library(DOSE)

geneList &lt;- as.vector(Allgenes_limma_pw$logFC)
names(geneList) &lt;- rownames(Allgenes_limma_pw)
gene &lt;- na.omit(DEgenes_pw$ENTREZID)

OrgDb &lt;- org.Hs.eg.db::org.Hs.eg.db # can also be org.Mm.eg.db::org.Mm.eg.db

# Group GO
ggo &lt;- clusterProfiler::groupGO(gene     = gene,
                                OrgDb    = OrgDb,
                                ont      = &quot;BP&quot;,
                                level    = 3,
                                readable = TRUE)
head(summary(ggo))
barplot(ggo, drop=TRUE, showCategory=12)

# GO over-representation test
ego &lt;- clusterProfiler::enrichGO(gene          = gene,
                                 OrgDb         = OrgDb,
                                 ont           = &quot;BP&quot;,
                                 pAdjustMethod = &quot;BH&quot;,
                                 pvalueCutoff  = 0.05,
                                 qvalueCutoff  = 0.05, 
                                 readable      = TRUE)
head(summary(ego))
barplot(ego, showCategory=25)
clusterProfiler::dotplot(ego, showCategory=25)
#enrichMap(ego)
cnetplot(ego, categorySize=&quot;pvalue&quot;, foldChange=geneList)
clusterProfiler::plotGOgraph(ego)

# enrichGO test the whole GO corpus and enriched result may contains very general terms. With dropGO function, user can remove specific GO terms or GO level from results obtained from both enrichGO and compareCluster.


## KEGG over-representation test
kk &lt;- clusterProfiler::enrichKEGG(gene         = gene,
                 organism     = 'hsa',
                 pvalueCutoff = 0.05)
head(summary(kk))
barplot(kk, showCategory=8)
clusterProfiler::dotplot(kk)
cnetplot(kk, categorySize=&quot;geneNum&quot;, foldChange=geneList)
</code></pre>

<hr />

<p><br></p>

<h2 id="pathview">Pathview</h2>

<p>uses the R package pathview.</p>

<p>For Toll-like-receptor signaling:</p>

<pre><code class="language-r">pathview_func(DEgenes_pw, logFCcolumn=&quot;logFC&quot;, pathway.id = &quot;04620&quot;, out.suffix = &quot;DE_TolLPS&quot;)
</code></pre>

<hr />

<p><br></p>

<h2 id="venn-diagrams-and-biological-theme-comparison">Venn diagrams and Biological theme comparison</h2>

<pre><code class="language-r">library(gplots)

venn_list &lt;- list(genelist1 = na.omit(rownames(DEgenes_pw)[1:50]), genelist2 = na.omit(rownames(DEgenes_pw)[30:80]), genelist3 = na.omit(rownames(DEgenes_pw)[48:100]))

gplots::venn(venn_list, show.plot=TRUE, small=0.7, showSetLogicLabel=FALSE, simplify = TRUE)

mtext(&quot;If you want a header, put it here&quot;, side=3, cex = 0.8)
</code></pre>

<pre><code class="language-r"># Biological theme comparison
compGO &lt;- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                          fun           = &quot;enrichGO&quot;,
                                          ont           = &quot;BP&quot;,
                                          OrgDb         = OrgDb,
                                          qvalueCutoff  = 0.001,
                                          pAdjustMethod = &quot;BH&quot;,
                                          readable      = TRUE)


compKEGG &lt;- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                            fun           = &quot;enrichKEGG&quot;,
                                            qvalueCutoff  = 0.05,
                                            pAdjustMethod = &quot;BH&quot;)
compKEGG &lt;- clusterProfiler::setReadable(compKEGG, OrgDb = OrgDb)


compMKEGG &lt;- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                             fun           = &quot;enrichMKEGG&quot;,
                                             organism='hsa', minGSSize=1)

compDO &lt;- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                          fun           = &quot;enrichDO&quot;,
                                          qvalueCutoff  = 0.05,
                                          pAdjustMethod = &quot;BH&quot;,
                                          readable      = TRUE)

# For visualisation see Enrichment Analysis using clusterProfiler
</code></pre>

<hr />

<p><br></p>

<h1 id="networks">Networks</h1>

<h2 id="tf-networks">TF networks</h2>

<p>Organism can be human or mouse. For human, the longer published list of TFs is used; for mouse the shorter list provided by Bonn (for which I don&rsquo;t have any more info on where it comes from).</p>

<p>This function produced two output files:</p>

<ul>
<li>The .expression matrix of TF expression data.</li>
</ul>

<p>It will have to be opened with <strong>Biolayout3D</strong>: Set Minimum Correlation and Correlation metric (by default 0.7 and Pearson). Then choose a suitable correlation coefficient (Graph Degree Distribution should be close to linear). Save the resulting network as a TGF file.</p>

<p>Open the TGF file in <strong>Cytoscape</strong>: Open network from file. Then got to Advanced Options: untick &ldquo;Use first column names&rdquo; and add &ldquo; to &ldquo;Other:&ldquo;. Now set Column 2 as Source and Column 5 as Target.</p>

<ul>
<li>Open the node annotation table in .txt format.</li>
</ul>

<p>In cytoscape, open table from file, import data as Node Table Columns. You can then customize the look of your network.</p>

<p>For example, go to Style and</p>

<ul>
<li><p>go to Border Paint, Column: DE, Mapping type: Discrete and choose a color for 1 and 0</p></li>

<li><p>go to Fill Color, Column logFC, Mapping Type: Continuous and choose colors.</p></li>
</ul>

<pre><code class="language-r">TF_networks(expmatrix, nodeAnno=Allgenes_limma_pw)
</code></pre>

<p><img src="CopyOfnetwork_test.png" alt="TF network Pearson 0.7" /></p>

<hr />

<p><br></p>

<h2 id="bingo-cytoscape">BiNGO (Cytoscape)</h2>

<ul>
<li><p>Go to Cytoscape</p></li>

<li><p>Apps -&gt; <strong>BiNGO</strong></p></li>

<li><p>The BiNGO Settings panel pops up. Start by filling in a name for your cluster. This name will be used for creating the output file and the visualization of the results in Cytoscape.</p></li>

<li><p>Check Paste Genes from Text and paste gene names into field below</p></li>

<li><p>We want to assess overrepresentation of GO categories, and we want to visualize the results in Cytoscape. The corresponding boxes are checked accordingly by default. Then select a statistical test (the Hypergeometric Test is exact and equivalent to an exact Fisher test, the Binomial Test is less accurate but quicker) and a multiple testing correction (we recommend Benjamini &amp; Hochberg&rsquo;s FDR correction, the Bonferroni correction will be too conservative in most cases), and choose a significance level, e.g. 0.05. Since we only want to visualize those GO categories that are overrepresented after multiple testing correction, and their parents in the GO hierarchy, select the corresponding visualization option. We&rsquo;re interested in assessing the overrepresentation of functional categories in our cluster with respect to the whole genome, which is why we choose the Complete Annotation as the reference set.</p></li>

<li><p>Select GO _Biological _Process from the ontology list, and Human from the organism list. We want to consider all evidence codes, so don&rsquo;t fill in anything in the evidence code box.</p></li>

<li><p>Finally, select a directory to save the output file in (the file will be named test.bgo if you filled in test as a cluster name), and press Start BiNGO&hellip;</p></li>

<li><p>The program will inform you of its progress while parsing the annotations and calculating the tests, corrections and layout. Finally, a visualization of the overrepresented GO categories is created in Cytoscape. Uncolored nodes are not overrepresented, but they are the parents of overrepresented categories further down. Yellow nodes represent GO categories that are overrepresented at the significance level . For more significant p-values, the node color gets increasingly more orange (see also the Color Legend panel). If you&rsquo;d like another layout, e.g. hierarchical, select the corresponding option from the Cytoscape Visualization menu. Regardless of the layout you choose, you&rsquo;ll probably have to tweak the nodes a little in order to avoid overlapping node labels. The list of significantly overrepresented functional categories is shown in the BiNGO output window (more information on the cluster and options you selected, and on which genes did not produce any annotation, is stored in the test.bgo file).</p></li>
</ul>

<hr />

<p><br></p>

<h1 id="other-functions">Other functions</h1>

<h2 id="mirna-target-identification">miRNA target identification</h2>

<pre><code class="language-r"># The current version of multiMiR is 1.0.1 and can be downloaded from
http://multimir.ucdenver.edu/multiMiR_1.0.1.tar.gz.

install.packages(&quot;XML&quot;)
install.packages(&quot;RCurl&quot;)
install.packages(&quot;/pathname/multiMiR_1.0.1.tar.gz&quot;, repos=NULL, type=&quot;source&quot;)

library(multiMiR)

miRNA_list &lt;- as.character(c(&quot;hsa-miR-146b-3p&quot;, &quot;hsa-miR-155-5p&quot;, &quot;hsa-miR-4521&quot;))

miRNA_allTargets = get.multimir(org=&quot;hsa&quot;, mirna=miRNA_list, target=NULL, table=&quot;all&quot;, summary=TRUE, predicted.cutoff.type=&quot;p&quot;, predicted.cutoff=NULL)
str(miRNA_allTargets)
    
write.table(miRNA_allTargets$predicted, &quot;miRNAs_allTargets_predictedTargets.txt&quot;, row.names = F, col.names = T, sep=&quot;\t&quot;)
write.table(miRNA_allTargets$validated, &quot;miRNAs_allTargets_validatedTargets.txt&quot;, row.names = F, col.names = T, sep=&quot;\t&quot;)
write.table(miRNA_allTargets$disease.drug, &quot;miRNAs_allTargets_diseaseDrugTargets.txt&quot;, row.names = F, col.names = T, sep=&quot;\t&quot;)
write.table(miRNA_allTargets$summary, &quot;miRNAs_allTargets_summaryTargets.txt&quot;, row.names = F, col.names = T, sep=&quot;\t&quot;)
</code></pre>

<hr />

<p><br></p>

<h2 id="hypergeometric-test-for-enrichment">Hypergeometric test for enrichment</h2>

<p>Analogous to Fisher&rsquo;s Exact Test.</p>

<pre><code class="language-r">phyper(SampleSuccesses -1, PopulationSuccesses, PopulationSize - PopulationSuccesses, sampleSize, lower.tail = FALSE)
</code></pre>

<hr />

<p><br></p>

<h1 id="citations">Citations</h1>

<blockquote>
<p>Martin Morgan, Herve Pages, Valerie Obenchain and Nathaniel Hayden (2016). Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import. R package version 1.24.0. <a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">http://bioconductor.org/packages/release/bioc/html/Rsamtools.html</a></p>

<p>Lawrence M, Huber W, Pages H, Aboyoun P, Carlson M, et al. (2013) Software for Computing and Annotating Genomic Ranges. PLoS Comput Biol 9(8): e1003118. <a href="doi:10.1371/journal.pcbi.1003118">doi:10.1371/journal.pcbi.1003118</a></p>

<p>Martin Morgan, Valerie Obenchain, Michel Lang and Ryan Thompson (2016). BiocParallel: Bioconductor facilities for parallel evaluation. R package version 1.6.2. <a href="https://github.com/Bioconductor/BiocParallel">https://github.com/Bioconductor/BiocParallel</a></p>

<p>Martin Morgan, Valerie Obenchain, Jim Hester and Herve Pages (2016). SummarizedExperiment: SummarizedExperiment container. R package version 1.2.2.</p>

<p>Michael I Love, Wolfgang Huber and Simon Anders (2014): Moderated estimation of fold change and dispersion for RNA-Seq data with DESeq2. Genome Biology</p>

<p>Gautier, L., Cope, L., Bolstad, B. M., and Irizarry, R. A. 2004. affy&mdash;analysis of Affymetrix GeneChip data at the probe level. Bioinformatics 20, 3 (Feb. 2004), 307-315.</p>

<p>Carvalho B. S., and Irizarry, R. A. 2010. A Framework for Oligonucleotide Microarray Preprocessing. Bioinformatics.</p>

<p>Kauffmann, A., Gentleman, R.,, Huber, W. (2009) arrayQualityMetrics&ndash;a bioconductor package for quality assessment of microarray data. Bioinformatics, 25(3):415-6.</p>

<p>Dunning, M.J., Smith, M.L., Ritchie, M.E., Tavare, S. beadarray: R classes and methods for Illumina bead-based data, Bioinformatics. 2007, 23(16):2183-4.</p>

<p>Jeffrey T. Leek, W. Evan Johnson, Hilary S. Parker, Elana J. Fertig, Andrew E. Jaffe and John D. Storey (2016). sva: Surrogate Variable Analysis. R package version 3.20.0.</p>

<p>Henrik Bengtsson (2016). matrixStats: Functions that Apply to Rows and Columns of Matrices (and to Vectors). R package version 0.50.2. <a href="https://CRAN.R-project.org/package=matrixStats">https://CRAN.R-project.org/package=matrixStats</a></p>

<p>Kevin Wright (2015). corrgram: Plot a Correlogram. R package version 1.8. <a href="https://CRAN.R-project.org/package=corrgram">https://CRAN.R-project.org/package=corrgram</a></p>

<p>Hansen M, Gerds TA, Nielsen OH, Seidelin JB, Troelsen JT, et al. (2012) pcaGoPromoter - An R Package for Biological and Regulatory Interpretation of Principal Components in Genome-Wide Gene Expression Data. PLoS ONE 7(2): e32394. <a href="doi:10.1371/journal.pone.0032394">doi:10.1371/journal.pone.0032394</a></p>

<p>Raivo Kolde (2015). pheatmap: Pretty Heatmaps. R package version 1.0.8. <a href="https://CRAN.R-project.org/package=pheatmap">https://CRAN.R-project.org/package=pheatmap</a></p>

<p>Gregory R. Warnes, Ben Bolker, Lodewijk Bonebakker, Robert Gentleman, Wolfgang Huber Andy Liaw, Thomas Lumley, Martin Maechler, Arni Magnusson, Steffen Moeller, Marc Schwartz and Bill Venables (2016). gplots: Various R Programming Tools for Plotting Data. R package version 3.0.1. <a href="https://CRAN.R-project.org/package=gplots">https://CRAN.R-project.org/package=gplots</a></p>

<p>Tal Galili (2016). heatmaply: Interactive Heat Maps Using &lsquo;plotly&rsquo;. R package version 0.3.2. <a href="https://CRAN.R-project.org/package=heatmaply">https://CRAN.R-project.org/package=heatmaply</a></p>

<p>Langfelder P and Horvath S, WGCNA: an R package for weighted correlation network analysis. BMC Bioinformatics 2008, 9:559 <a href="doi:10.1186/1471-2105-9-559">doi:10.1186/1471-2105-9-559</a></p>

<p>Peter Langfelder, Steve Horvath (2012). Fast R Functions for Robust Correlations and Hierarchical Clustering. Journal of Statistical Software, 46(11), 1-17. URL <a href="http://www.jstatsoft.org/v46/i11/">http://www.jstatsoft.org/v46/i11/</a>.</p>

<p>Herve Pages, Marc Carlson, Seth Falcon and Nianhua Li (2016). AnnotationDbi: Annotation Database Interface. R package version 1.34.3.</p>

<p>Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47.</p>

<p>David B. Dahl (2016). xtable: Export Tables to LaTeX or HTML. R package version 1.8-2. <a href="https://CRAN.R-project.org/package=xtable">https://CRAN.R-project.org/package=xtable</a></p>

<p>H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2009.</p>

<p>Kamil Slowikowski (2016). ggrepel: Repulsive Text and Label Geoms for &lsquo;ggplot2&rsquo;. R package version 0.5. <a href="https://CRAN.R-project.org/package=ggrepel">https://CRAN.R-project.org/package=ggrepel</a></p>

<p>Guangchuang Yu, Li-Gen Wang, Yanyan Han and Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology 2012, 16(5):284-287</p>

<p>Guangchuang Yu, Li-Gen Wang, Guang-Rong Yan, Qing-Yu He. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics 2015 31(4):608-609</p>

<p>Marc Carlson (2016). org.Hs.eg.db: Genome wide annotation for Human. R package version 3.3.0.</p>

<p>Marc Carlson (2016). org.Mm.eg.db: Genome wide annotation for Mouse. R package version 3.3.0.</p>

<p>Luo, W. and Brouwer C., Pathview: an R/Bioconductor package for pathway-based data integration and visualization. Bioinformatics, 2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285</p>

<p>Yuanbin Ru and Katerina Kechris (2014). multiMiR: Integration of multiple microRNA-target databases with their disease and drug associations. R package version 1.0.1.</p>
</blockquote>

<h1 id="session-info">Session Info</h1>

<pre><code class="language-r">sessionInfo()
</code></pre>

<pre><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 7 x64 (build 7601) Service Pack 1
## 
## locale:
## [1] LC_COLLATE=English_United States.1252 
## [2] LC_CTYPE=English_United States.1252   
## [3] LC_MONETARY=English_United States.1252
## [4] LC_NUMERIC=C                          
## [5] LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] exprAnalysis_0.1.0
## 
## loaded via a namespace (and not attached):
##  [1] fastcluster_1.1.20    gtools_3.5.0          splines_3.3.1        
##  [4] lattice_0.20-34       colorspace_1.2-6      htmltools_0.3.5      
##  [7] stats4_3.3.1          yaml_2.1.13           chron_2.3-47         
## [10] survival_2.39-5       pcaGoPromoter_1.16.0  foreign_0.8-66       
## [13] DBI_0.5               BiocGenerics_0.18.0   RColorBrewer_1.1-2   
## [16] matrixStats_0.50.2    foreach_1.4.3         plyr_1.8.4           
## [19] stringr_1.1.0         zlibbioc_1.18.0       Biostrings_2.40.2    
## [22] munsell_0.4.3         gtable_0.2.0          caTools_1.17.1       
## [25] codetools_0.2-14      evaluate_0.9          latticeExtra_0.6-28  
## [28] Biobase_2.32.0        knitr_1.14            IRanges_2.6.1        
## [31] doParallel_1.0.10     parallel_3.3.1        AnnotationDbi_1.34.4 
## [34] preprocessCore_1.34.0 Rcpp_0.12.7           acepack_1.3-3.3      
## [37] KernSmooth_2.23-15    flashClust_1.01-2     scales_0.4.0         
## [40] formatR_1.4           gdata_2.17.0          org.Hs.eg.db_3.3.0   
## [43] S4Vectors_0.10.3      Hmisc_3.17-4          WGCNA_1.51           
## [46] XVector_0.12.1        gplots_3.0.1          gridExtra_2.2.1      
## [49] impute_1.46.0         ellipse_0.3-8         ggplot2_2.1.0        
## [52] digest_0.6.10         stringi_1.1.1         grid_3.3.1           
## [55] tools_3.3.1           bitops_1.0-6          magrittr_1.5         
## [58] RSQLite_1.0.0         dynamicTreeCut_1.63-1 Formula_1.2-1        
## [61] cluster_2.0.4         GO.db_3.3.0           Matrix_1.2-7.1       
## [64] data.table_1.9.6      rmarkdown_1.0         iterators_1.0.8      
## [67] rpart_4.1-10          nnet_7.3-12
</code></pre>

              
            </div>
          </div>
          <div id="post-footer" class="post-footer main-content-wrap">
            
              
                
                
                  <div class="post-footer-tags">
                    <span class="text-color-light text-small">TAGGED IN</span><br/>
                    
  <a class="tag tag--primary tag--small" href="//tags/rna-seq/">RNA-seq</a>

  <a class="tag tag--primary tag--small" href="//tags/expranalysis/">exprAnalysis</a>

  <a class="tag tag--primary tag--small" href="//tags/illumina/">Illumina</a>

  <a class="tag tag--primary tag--small" href="//tags/affymetrix/">Affymetrix</a>

  <a class="tag tag--primary tag--small" href="//tags/deseq2/">DESeq2</a>

  <a class="tag tag--primary tag--small" href="//tags/ggplot2/">ggplot2</a>

  <a class="tag tag--primary tag--small" href="//tags/networks/">networks</a>

  <a class="tag tag--primary tag--small" href="//tags/cummerbund/">cummeRbund</a>

  <a class="tag tag--primary tag--small" href="//tags/pca/">PCA</a>

  <a class="tag tag--primary tag--small" href="//tags/heatmap/">heatmap</a>

  <a class="tag tag--primary tag--small" href="//tags/wgcna/">WGCNA</a>

                  </div>
                
              
            
            
  <div class="post-actions-wrap">
      <nav >
        <ul class="post-actions post-action-nav">
            <li class="post-action">
              
                <a class="post-action-btn btn btn--default tooltip--top" href="/2017/05/data-science-business-forecasting-1/" data-tooltip="Data Science for Business - Time Series Forecasting Part 1: EDA &amp; Data Preparation">
              
                <i class="fa fa-angle-left"></i>
                <span class="hide-xs hide-sm text-small icon-ml">NEXT</span>
              </a>
            </li>
            <li class="post-action">
              
                <a class="post-action-btn btn btn--default tooltip--top" href="/2013/02/image-gallery-showcase/" data-tooltip="Image gallery showcase">
              
                <span class="hide-xs hide-sm text-small icon-mr">PREVIOUS</span>
                <i class="fa fa-angle-right"></i>
              </a>
            </li>
        </ul>
      </nav>
    <ul class="post-actions post-action-share" >
      
        <li class="post-action hide-lg hide-md hide-sm">
          <a class="post-action-btn btn btn--default btn-open-shareoptions" href="#btn-open-shareoptions">
            <i class="fa fa-share-alt"></i>
          </a>
        </li>
        
          <li class="post-action hide-xs">
            <a class="post-action-btn btn btn--default" target="new" href="https://www.facebook.com/sharer/sharer.php?u=/2016/09/expranalyis-package/">
              <i class="fa fa-facebook-official"></i>
            </a>
          </li>
        
          <li class="post-action hide-xs">
            <a class="post-action-btn btn btn--default" target="new" href="https://twitter.com/intent/tweet?text=/2016/09/expranalyis-package/">
              <i class="fa fa-twitter"></i>
            </a>
          </li>
        
          <li class="post-action hide-xs">
            <a class="post-action-btn btn btn--default" target="new" href="https://plus.google.com/share?url=/2016/09/expranalyis-package/">
              <i class="fa fa-google-plus"></i>
            </a>
          </li>
        
      
      
        <li class="post-action">
          <a class="post-action-btn btn btn--default" href="#disqus_thread">
            <i class="fa fa-comment-o"></i>
          </a>
        </li>
      
      <li class="post-action">
        
          <a class="post-action-btn btn btn--default" href="#">
        
          <i class="fa fa-list"></i>
        </a>
      </li>
    </ul>
  </div>


            
              
                <div id="disqus_thread">
  <noscript>Please enable JavaScript to view the <a href="//disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</div>
              
            
          </div>
        </article>
        <footer id="footer" class="main-content-wrap">
  <span class="copyrights">
    &copy; 2017 Dr. Shirin Glander. All Rights Reserved
  </span>
</footer>

      </div>
      <div id="bottom-bar" class="post-bottom-bar" data-behavior="4">
        
  <div class="post-actions-wrap">
      <nav >
        <ul class="post-actions post-action-nav">
            <li class="post-action">
              
                <a class="post-action-btn btn btn--default tooltip--top" href="/2017/05/data-science-business-forecasting-1/" data-tooltip="Data Science for Business - Time Series Forecasting Part 1: EDA &amp; Data Preparation">
              
                <i class="fa fa-angle-left"></i>
                <span class="hide-xs hide-sm text-small icon-ml">NEXT</span>
              </a>
            </li>
            <li class="post-action">
              
                <a class="post-action-btn btn btn--default tooltip--top" href="/2013/02/image-gallery-showcase/" data-tooltip="Image gallery showcase">
              
                <span class="hide-xs hide-sm text-small icon-mr">PREVIOUS</span>
                <i class="fa fa-angle-right"></i>
              </a>
            </li>
        </ul>
      </nav>
    <ul class="post-actions post-action-share" >
      
        <li class="post-action hide-lg hide-md hide-sm">
          <a class="post-action-btn btn btn--default btn-open-shareoptions" href="#btn-open-shareoptions">
            <i class="fa fa-share-alt"></i>
          </a>
        </li>
        
          <li class="post-action hide-xs">
            <a class="post-action-btn btn btn--default" target="new" href="https://www.facebook.com/sharer/sharer.php?u=/2016/09/expranalyis-package/">
              <i class="fa fa-facebook-official"></i>
            </a>
          </li>
        
          <li class="post-action hide-xs">
            <a class="post-action-btn btn btn--default" target="new" href="https://twitter.com/intent/tweet?text=/2016/09/expranalyis-package/">
              <i class="fa fa-twitter"></i>
            </a>
          </li>
        
          <li class="post-action hide-xs">
            <a class="post-action-btn btn btn--default" target="new" href="https://plus.google.com/share?url=/2016/09/expranalyis-package/">
              <i class="fa fa-google-plus"></i>
            </a>
          </li>
        
      
      
        <li class="post-action">
          <a class="post-action-btn btn btn--default" href="#disqus_thread">
            <i class="fa fa-comment-o"></i>
          </a>
        </li>
      
      <li class="post-action">
        
          <a class="post-action-btn btn btn--default" href="#">
        
          <i class="fa fa-list"></i>
        </a>
      </li>
    </ul>
  </div>


      </div>
      <div id="share-options-bar" class="share-options-bar" data-behavior="4">
  <i id="btn-close-shareoptions" class="fa fa-close"></i>
  <ul class="share-options">
    
      <li class="share-option">
        <a class="share-option-btn" target="new" href="https://www.facebook.com/sharer/sharer.php?u=%2F2016%2F09%2Fexpranalyis-package%2F">
          <i class="fa fa-facebook-official"></i><span>Share on Facebook</span>
        </a>
      </li>
    
      <li class="share-option">
        <a class="share-option-btn" target="new" href="https://twitter.com/intent/tweet?text=%2F2016%2F09%2Fexpranalyis-package%2F">
          <i class="fa fa-twitter"></i><span>Share on Twitter</span>
        </a>
      </li>
    
      <li class="share-option">
        <a class="share-option-btn" target="new" href="https://plus.google.com/share?url=%2F2016%2F09%2Fexpranalyis-package%2F">
          <i class="fa fa-google-plus"></i><span>Share on Google&#43;</span>
        </a>
      </li>
    
  </ul>
</div>
<div id="share-options-mask" class="share-options-mask"></div>
    </div>
    
    <div id="about">
  <div id="about-card">
    <div id="about-btn-close">
      <i class="fa fa-remove"></i>
    </div>
    
      <img id="about-card-picture" src="//www.gravatar.com/avatar/7f7f818e55624edfef8aa93860a1a3d4?s=110" alt="Author&#39;s picture" />
    
    <h4 id="about-card-name">Dr. Shirin Glander</h4>
    
      <div id="about-card-bio">Biologist turned Bioinformatician turned Data Scientist</div>
    
    
      <div id="about-card-job">
        <i class="fa fa-briefcase"></i>
        <br/>
        Data Scientist
      </div>
    
    
      <div id="about-card-location">
        <i class="fa fa-map-marker"></i>
        <br/>
        Münster, Germany
      </div>
    
  </div>
</div>

    <div id="algolia-search-modal" class="modal-container">
  <div class="modal">
    <div class="modal-header">
      <span class="close-button"><i class="fa fa-close"></i></span>
      <a href="https://algolia.com" target="_blank" rel="noopener" class="searchby-algolia text-color-light link-unstyled">
        <span class="searchby-algolia-text text-color-light text-small">by</span>
        <img class="searchby-algolia-logo" src="https://www.algolia.com/static_assets/images/press/downloads/algolia-light.svg">
      </a>
      <i class="search-icon fa fa-search"></i>
      <form id="algolia-search-form">
        <input type="text" id="algolia-search-input" name="search"
          class="form-control input--large search-input" placeholder="Search" />
      </form>
    </div>
    <div class="modal-body">
      <div class="no-result text-color-light text-center">no post found</div>
      <div class="results">
        
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/09/moving-my-blog-to-blogdown/">
                <h3 class="media-heading">Moving my blog to blogdown</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Sep 9, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">It’s been a long time coming but I finally finished moving my blog from Jekyll and Bootstrap to blogdown and Hugo! Yay!!!
Just to test that everything works, I run the example code:
R Markdown This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see http://rmarkdown.rstudio.com.
You can embed an R code chunk like this:</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/09/data-science-fraud-detection/">
                <h3 class="media-heading">Data Science for Fraud Detection</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Sep 9, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">I have written the following post about Data Science for Fraud Detection at my company codecentric&rsquo;s blog:
 Fraud can be defined as “the crime of getting money by deceiving people” (Cambridge Dictionary); it is as old as humanity: whenever two parties exchange goods or conduct business there is the potential for one party scamming the other. With an ever increasing use of the internet for shopping, banking, filing insurance claims, etc.</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/09/migrating-github-gitlab/">
                <h3 class="media-heading">Migrating from GitHub to GitLab with RStudio (Tutorial)</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Sep 9, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">GitHub vs. GitLab Git is a distributed implementation of version control. Many people have written very eloquently about why it is a good idea to use version control, not only if you collaborate in a team but also if you work on your own; one example is this article from RStudio&rsquo;s Support pages.
In short, its main feature is that version control allows you to keep track of the changes you make to your code.</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/07/twitter-analysis-codecentric/">
                <h3 class="media-heading">Social Network Analysis and Topic Modeling of codecentric’s Twitter friends and followers</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Jul 7, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">I have written the following post about Social Network Analysis and Topic Modeling of codecentric&rsquo;s Twitter friends and followers for codecentric&rsquo;s blog:
 Recently, Matthias Radtke has written a very nice blog post on Topic Modeling of the codecentric Blog Articles, where he is giving a comprehensive introduction to Topic Modeling. In this article I am showing a real-world example of how we can use Data Science to gain insights from text data and social network analysis.</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/07/ocr_tesseract/">
                <h3 class="media-heading">How to do Optical Character Recognition (OCR) of non-English documents in R using Tesseract?</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Jul 7, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">One of the many great packages of rOpenSci has implemented the open source engine Tesseract.
Optical character recognition (OCR) is used to digitize written or typed documents, i.e. photos or scans of text documents are &ldquo;translated&rdquo; into a digital text on your computer.
While this might seem like a trivial task at first glance, because it is so easy for our human brains. When reading text, we make use of our built-in word and sentence &ldquo;autocomplete&rdquo; that we learned from experience.</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/06/data-science-business-forecasting-3/">
                <h3 class="media-heading">Data Science for Business - Time Series Forecasting Part 3: Forecasting with Facebook&#39;s Prophet</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Jun 6, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">In my last two posts (Part 1 and Part 2), I explored time series forecasting with the timekit package.
In this post, I want to compare how Facebook&rsquo;s prophet performs on the same dataset.

Predicting future events/sales/etc. isn&rsquo;t trivial for a number of reasons and different algorithms use different approaches to handle these problems. Time series data does not behave like a regular numeric vector, because months don&rsquo;t have the same number of days, weekends and holidays differ between years, etc.</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/06/data-science-business-forecasting-2/">
                <h3 class="media-heading">Data Science for Business - Time Series Forecasting Part 2: Forecasting with timekit</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Jun 6, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">In my last post, I prepared and visually explored time series data.
Now, I will use this data to test the timekit package for time series forecasting with machine learning.

Forecasting In time series forecasting, we use models to predict future time points based on past observations.
As mentioned in timekit&rsquo;s vignette, &ldquo;as with most machine learning applications, the prediction is only as good as the patterns in the data.</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2017/05/data-science-business-forecasting-1/">
                <h3 class="media-heading">Data Science for Business - Time Series Forecasting Part 1: EDA &amp; Data Preparation</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  May 5, 2017
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">Data Science is a fairly broad term and encompasses a wide range of techniques from data visualization to statistics and machine learning models. But the techniques are only tools in a - sometimes very messy - toolbox. And while it is important to know and understand these tools, here, I want to go at it from a different angle: What is the task at hand that data science tools can help tackle, and what question do we want to have answered?</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2016/09/expranalyis-package/">
                <h3 class="media-heading">exprAnalysis package</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Sep 9, 2016
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather">I created the R package exprAnalysis designed to streamline my RNA-seq data analysis pipeline. Below you find the vignette for installation and usage of the package.

This package combines functions from various packages used to analyze and visualize expression data from NGS or expression chips.
 It supports normalized input as e.g. from Cufflinks or expression chip arrays and raw count data from bam file input. It supports mRNA, miRNA, protein or other expression data.</div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
          <div class="media">
            
            <div class="media-body">
              <a class="link-unstyled" href="/2013/02/image-gallery-showcase/">
                <h3 class="media-heading">Image gallery showcase</h3>
              </a>
              <span class="media-meta">
                <span class="media-date text-small">
                  Feb 2, 2013
                </span>
              </span>
              <div class="media-content hide-xs font-merryweather"><p>And here is the image gallery!
</p></div>
            </div>
            <div style="clear:both;"></div>
            <hr>
          </div>
        
      </div>
    </div>
    <div class="modal-footer">
      <p class="results-count text-medium"
         data-message-zero="no post found"
         data-message-one="1 post found"
         data-message-other="{n} posts found">
         11 posts found
      </p>
    </div>
  </div>
</div>
    
  
    
    <div id="cover" style="background-image:url('/images/cover.jpg');"></div>
  


    
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.3/jquery.min.js" integrity="sha256-IFHWFEbU2/+wNycDECKgjIRSirRNIDp2acEB5fvdVRU=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.8.0/highlight.min.js" integrity="sha256-+mpyNVJsNt4rVXCw0F+pAOiB3YxmHgrbJsx4ecPuUaI=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/2.1.4/jquery.fancybox.min.js" integrity="sha256-vMxgR/7FtLovVA+IPrR7+xTgIgARH7y9VZQnmmi0HDI=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/2.1.4/helpers/jquery.fancybox-thumbs.min.js" integrity="sha256-N0qFUh7/9vLvia87dDndewmsgsyYoNkdA212tPc+2NI=" crossorigin="anonymous"></script>


<script src="/js/script-qi9wbxp2ya2j6p7wx1i6tgavftewndznf4v0hy2gvivk1rxgc3lm7njqb6bz.min.js"></script>


<script>
$(document).ready(function() {
  hljs.configure({ classPrefix: '', useBR: false });
  $('pre.code-highlight, pre > code').each(function(i, block) {
    if (!$(this).hasClass('codeblock')) {
      $(this).addClass('codeblock');
    }
    hljs.highlightBlock(block);
  });
});
</script>

  
    
      <script>
        var disqus_config = function () {
          this.page.url = '\/2016\/09\/expranalyis-package\/';
          
            this.page.identifier = '\/2016\/09\/expranalyis-package\/'
          
        };
        (function() {
          
          
          if (window.location.hostname == "localhost") {
            return;
          }
          var d = document, s = d.createElement('script');
          var disqus_shortname = 'shirinsplayground';
          s.src = '//' + disqus_shortname + '.disqus.com/embed.js';

          s.setAttribute('data-timestamp', +new Date());
          (d.head || d.body).appendChild(s);
        })();
      </script>
    
  



    
  </body>
</html>

